src/main/java/org/broadinstitute/hellbender/engine/ReferenceDataSourceFromFile.java

```diff
@@ -0,0 +1,120 @@
+package org.broadinstitute.hellbender.engine;
+
+import htsjdk.samtools.SAMSequenceDictionary;
+import htsjdk.samtools.reference.ReferenceSequence;
+import org.broadinstitute.hellbender.exceptions.GATKException;
+import org.broadinstitute.hellbender.utils.SimpleInterval;
+import org.broadinstitute.hellbender.utils.fasta.CachingIndexedFastaSequenceFile;
+import org.broadinstitute.hellbender.utils.iterators.ByteArrayIterator;
+
+import java.io.File;
+import java.io.IOException;
+import java.util.Iterator;
+
+/**
+ * Manages traversals and queries over reference data (for now, fasta files only)
+ *
+ * Supports targeted queries over the reference by interval, but does not
+ * yet support complete iteration over the entire reference.
+ */
+public final class ReferenceDataSourceFromFile implements ReferenceDataSource {
+
+    /**
+     * Our reference file. Uses the caching version of IndexedFastaSequenceFile
+     * so that repeated queries over nearby locations will be efficient (this
+     * is the primary reference access pattern in most traversals).
+     */
+    private CachingIndexedFastaSequenceFile reference;
+
+    /**
+     * Initialize this data source using a fasta file.
+     *
+     * The provided fasta file must have companion .fai and .dict files.
+     *
+     * @param fastaFile reference fasta file
+     */
+    public ReferenceDataSourceFromFile(final File fastaFile) {
+        if ( fastaFile == null )
+            throw new IllegalArgumentException("fastaFile must be non-null");
+
+        // Will throw a UserException if the .fai and/or .dict are missing
+        reference = CachingIndexedFastaSequenceFile.checkAndCreate(fastaFile);
+    }
+
+    /**
+     * Start an iteration over the entire reference. Not yet supported!
+     *
+     * See the BaseUtils class for guidance on how to work with bases in this format.
+     *
+     * @return iterator over all bases in this reference
+     */
+    @Override
+    public Iterator<Byte> iterator() {
+        throw new UnsupportedOperationException("Iteration over entire reference not yet implemented");
+    }
+
+    /**
+     * Query a specific interval on this reference, and get back an iterator over the bases spanning that interval.
+     *
+     * See the BaseUtils class for guidance on how to work with bases in this format.
+     *
+     * @param interval query interval
+     * @return iterator over the bases spanning the query interval
+     */
+    @Override
+    public Iterator<Byte> query( final SimpleInterval interval ) {
+        // TODO: need a way to iterate lazily over reference bases without necessarily loading them all into memory at once
+        return new ByteArrayIterator(queryAndPrefetch(interval).getBases());
+    }
```

Comment:
This implementation of `query()` should be replaced by a default implementation in the `ReferenceDataSource` interface that also returns `new ByteArrayIterator(queryAndPrefetch(interval).getBases());`